[flink-5568] [Table API & SQL]Introduce interface for catalog, and provide an in-memory implementation. Integrate external catalog with calcite catalog.

beyond1920 opened the pull request:
This pr aims to Introduce interface for catalog, and provide an in-memory implementation for test and develop,  finally integrate external catalog with calcite catalog.
The main change including:
1. Introduce FlinkExternalCatalog abstraction. Besides introduce ExternalCatalogDatabase as database in catalog  and ExternalCatalogTable as table in catalog.
2. Provide an in-memory implementation for test and develop.
3. Introduce FlinkExternalCatalogSchema which is an implementation of Calcite Schema interface. It registers database at FlinkExternalCatalog as calcite Schemas, and tables in a database as Calcite table.
4. Add ExternalCatalogCompatible annotation. The TableSource with this annotation represents it could be converted to or from externalCatalogTable.  ExternalCatalogTable2TableSourceConverter is the converter between externalCatalogTable and tableSource.

tonycox committed:
[FLINK-3849] [table] Add FilterableTableSource interface and rules for pushing it (1)

fix filterable test

rebase and trying fix rexnode parsing

create wrapper and update rules

jgrier committed:
[FLINK-5635] [docker] Improvements for Docker on Flink experience

Modifying Dockerfile to build from local flink-dist as well as release URLs.
Logging to stdout.
Adding scripts to deploy seamlessly on Docker Swarm.
Updating Docker Compose scripts to work correctly.
Parameterizing things so these Docker scripts are more generally useful.

tony810430 committed:
[FLINK-4754] [checkpoints] Make number of retained checkpoints user configurable

This closes #3374

jinmingjian committed:
[FLINK-5692] [core] Add an Option to Deactivate Kryo Fallback for Serializers

This closes #3373

jinmingjian committed:
[FLINK-4422] [clients] Convert time interval measurements to System.nanoTime() in 'flink-clients'

This closes #3384

godfreyhe commented on flink-libraries/flink-table/src/main/resources/externalcatalogTable.properties:
```diff
@@ -0,0 +1,19 @@
+# limitations under the License.
+################################################################################
+
+scan.package=org.apache.flink.table.sources
```
用户的 table 怎么定义 scan package

beyond1920 replied:
比如用户新写了一个tt的connector，包在com.taobao.blink***下，那么scan.package就是com.taobao.blink，会在这个package下寻找带有@ExternalCatalogCompatible注解的TableSource

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/ExternalCatalogTable.scala:
```diff
@@ -0,0 +1,70 @@
+    identifier: TableIdentifier,
+    tableType: String,
+    schema: DataSchema,
+    partitionColumnNames: Seq[String] = Seq.empty,
```
不是所有表都有 partition 的概念

beyond1920 replied:
done

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/ExternalCatalogTable.scala:
```diff
@@ -0,0 +1,70 @@
+  */
+case class TableIdentifier(
+    database: String,
+    table: String) {
```
不像 MySQL 这样的DB， CSV 这样的文件，database 是啥？ empty？ 所有的表都是三层结构：catalog_name -> db_name -> table_name ?

beyond1920 replied:
是的

beyond1920 replied:
一般是在DDL语句里面用到ExternalCatalog，根据规范，基本都是db.tb两层。如果有特殊的情况，一般会做一个faked的db层，比如default databse

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/ExternalCatalogTable2TableSourceConverter.scala:
```diff
@@ -0,0 +1,37 @@
+    */
+  def fromExternalCatalogTable(externalCatalogTable: ExternalCatalogTable): T
+
+}
```
toExternalCatalogTable 目前不涉及？

beyond1920 replied:
是的，暂时不需要

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalogSchema.scala:
```diff
@@ -0,0 +1,164 @@
+  * @param catalog           external catalog
+  */
+class FlinkExternalCatalogSchema(
+    CatalogIdentifier: String,
```
=> catalogIdentifier

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/ExternalCatalogTable2TableSourceConverter.scala:
```diff
@@ -0,0 +1,37 @@
+  *
+  * @tparam T The tableSource type which to do convert operation on.
+  */
+trait ExternalCatalogTable2TableSourceConverter[T <: TableSource[_]] {
```
是不是叫TableSourceConverter？

zhuoluoy replied:
Please rename. Do not use number "2" in a trait. Please use noun or adjectives. TableSourceConverter may be a good idea.

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/FlinkCatalogTableHelper.scala:
```diff
@@ -0,0 +1,138 @@
+/**
+  * The utility class is used to convert ExternalCatalogTable to TableSourceTable.
+  */
+object FlinkCatalogTableHelper {
```
FlinkCatalogTableHelper =》CatalogTableHelper

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,108 @@
+  * <li> provide tables for calcite catalog, it looks up databases or tables in the external catalog
+  * </ul>
+  */
+trait FlinkExternalCatalog {
```
FlinkExternalCatalog =》 ExternalCatalog

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalogSchema.scala:
```diff
@@ -0,0 +1,164 @@
+  * @param CatalogIdentifier external catalog name
+  * @param catalog           external catalog
+  */
+class FlinkExternalCatalogSchema(
```
FlinkExternalCatalogSchema =》 ExternalCatalogSchema

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+  *
+  * It could be used for testing or developing instead of used in production environment.
+  */
+class InMemoryFlinkExternalCatalog extends FlinkExternalCatalog {
```
InMemoryFlinkExternalCatalog  =》 InMemoryExternalCatalog

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSourceConverter.scala:
```diff
@@ -0,0 +1,65 @@
+      case Some(lenientStr) => lenientStr.toBoolean
+      case None => false
+    }
+    new CsvTableSource(
```
用这个source的builder？

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalogSchemaTest.scala:
```diff
@@ -0,0 +1,90 @@
+import org.apache.flink.table.utils.CommonTestData
+import org.junit.{Assert, Before, Test}
+
+class FlinkExternalCatalogSchemaTest {
```
FlinkExternalCatalogSchemaTest =》 ExternalCatalogSchemaTest

beyond1920 replied:
done

KurtYoung commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalogTest.scala:
```diff
@@ -0,0 +1,141 @@
+import org.apache.flink.table.api.{DatabaseAlreadyExistException, DatabaseNotExistException, TableAlreadyExistException, TableNotExistException}
+import org.junit.{Assert, Before, Test}
+
+class InMemoryFlinkExternalCatalogTest {
```
InMemoryFlinkExternalCatalogTest =》 InMemoryExternalCatalogTest

beyond1920 replied:
done

KurtYoung reviewed (commented):
把catalog包从plan里挪出来和它平级吧

zhuoluoy commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/ExternalCatalogTable.scala:
```diff
@@ -0,0 +1,70 @@
+  * @param table    table name
+  */
+case class TableIdentifier(
+    database: String,
```
Is there a database in Excel files?

beyond1920 replied:
一般是在DDL语句里面用到ExternalCatalog，根据规范，基本都是db.tb两层。如果有特殊的情况，一般会做一个faked的db层，比如default databse

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+
+  override def dropDatabase(
+      dbName: String,
+      ignoreIfNotExists: Boolean): Unit = synchronized {
```
同上

beyond1920 replied:
done

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+      ignoreIfNotExists: Boolean): Unit = synchronized {
+    val tables = getTables(dbName)
+    if (tables.remove(tableName).isEmpty) {
+      throw new TableNotExistException(dbName, tableName)
```
ignoreIfNotExists 为 true，不应该抛异常吧

beyond1920 replied:
ok

beyond1920 replied:
done

godfreyhe commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalogSchemaTest.scala:
```diff
@@ -0,0 +1,90 @@
+  def testGetSubSchema(): Unit = {
+    val allSchemaObjectNames = calciteCatalogReader
+        .getAllSchemaObjectNames(Lists.newArrayList(schemaName))
+    Assert.assertTrue(allSchemaObjectNames.size() == 2)
```
import org.junit.Assert._ ，不需要 Assert.

godfreyhe commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/catalog/FlinkExternalCatalogSchemaTest.scala:
```diff
@@ -0,0 +1,90 @@
+  def setUp(): Unit = {
+    val rootSchemaPlus: SchemaPlus = CalciteSchema.createRootSchema(true, false).plus()
+    val catalog = CommonTestData.getMockedFlinkExternalCatalog()
+    flinkCatalogSchema = FlinkExternalCatalogSchema.create(rootSchemaPlus, schemaName, catalog)
```
之后用户怎么方便加catalog？

beyond1920 replied:
在另外一个jira里面，https://issues.apache.org/jira/browse/FLINK-5570，  一会提个对应的pr。简单来说就是注册到tableEnvironment

zhuoluoy commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+
+  private val databases = new HashMap[String, Database]
+
+  override def createTable(
```
synchronized?

zhuoluoy commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+      tableName: String,
+      ignoreIfNotExists: Boolean): Unit = synchronized {
+    val tables = getTables(dbName)
+    if (tables.remove(tableName).isEmpty) {
```
synchronized?

beyond1920 replied:
想做个简单的线程安全的in-memory实现

zhuoluoy commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/catalog/InMemoryFlinkExternalCatalog.scala:
```diff
@@ -0,0 +1,129 @@
+    val tables = getTables(dbName)
+    val tableName = table.identifier.table
+    if (tables.contains(tableName)) {
+      tables.put(tableName, table)
```
synchronized?

Commit:
[FLINK-5134] [runtime] Aggregate ResourceSpec for chained operators when generating job graph

This closes #3455

StefanRRichter committed:
[FLINK-5715] Asynchronous snapshots for heap-based keyed state backend

WangTaoTheTonic committed:
[FLINK-5981] [security] Make SSL pick up configured protocols and cipher suites

This closes #3486

patricklucas committed:
[FLINK-5635] [docker] Use start-foreground in Docker entrypoint

docker-entrypoint.sh should error on invalid args

Improve docker build.sh cleanup

Dockerfile improvements per review

Remove unnecessary Dockerfile build steps

Now that docker-entrypoint.sh uses 'start-foreground', munging
flink-daemon.sh and overwriting the log4j config are not longer
necessary.

Improve Dockerfile and docker-entrypoint.sh

Clean up Dockerfile and improve docker-entrypoint.sh to support '--help'
and pass through non-(jobmanager|taskmanager) commands.

This closes #3205.
This closes #3494.

greghogan committed:
[FLINK-6009] [java api] Deprecate DataSetUtils#checksumHashCode

This is likely only used by Gelly and we have a more featureful
implementation allowing for multiple outputs and setting the job name.
Deprecation will allow this to be removed in Flink 2.0.

This closes #3516

kl0u committed:
[hotfix] [doc] Fix error in ProcessFunction example.

KurtYoung committed:
[FLINK-3849] [table] Add FilterableTableSource interface and rules for pushing it (2)

This closes #3520.

fix compilation failure

fix compilation failure again.

1. Deep copy TableSource when we copy TableSourceScan
2. unify push project into scan rule for both batch and stream

address comments.

expand project list before creating new RexProgram

update tests

aljoscha committed:
[FLINK-6001] Fix ContinuousEventTimeTrigger firing without state

kl0u committed:
[FLINK-6032] [cep] Clean-up operator state when not needed.

The CEP operator now cleans the registered state for a
key. This happens:
1) for the priority queue, when the queue is empty.
2) for the NFA, when its shared buffer is empty.
3) finally the key is removed from the watermark
   callback service if both the above are empty.

liuyuzhong7 committed:
[FLINK-6041] [streaming api] Move StreamingFunctionUtils to 'org.apache.flink.streaming.util'

This close #3532

StefanRRichter committed:
[FLINK-6044] Replace all unintentional calls to InputStream#read(...) with InputStream#readFully(...)

sunjincheng121 committed:
[FLINK-6018] Properly initialise StateDescriptor in AbstractStateBackend.getPartitionedState()

Commit:
[FLINK-6051] [metrics] [docs] Fix scope keys on Configuration page

This closes #3538.

StefanRRichter committed:
[FLINK-6061] [state backends] Throw IllegalStateException when using RocksDB keyed state with no key set

StephanEwen committed:
[FLINK-5692] [core] (followups) Add an Option to Deactivate Kryo Fallback for Serializers

StephanEwen committed:
[hotfix] [core] Fix/cleanup serialization test for ExecutionConfig

Commit:
[FLINK-6064][flip6] fix BlobServer connection in TaskExecutor

The hostname used for the BlobServer was set to the akka address which is
invalid for this use. Instead, this adds the hostname to the RpcGateway /
AkkaInvocationHandler so that this information is available to the TaskExecutor.

This closes #3551.

StephanEwen committed:
[FLINK-5962] [checkpoints] Remove scheduled cancel-task from timer queue to prevent memory leaks

This closes #3548

zentol committed:
[FLINK-5650] [py] Continuously check PyProcess health while waiting for incoming connection

StephanEwen committed:
[FLINK-5756] [rocksdb] Add mini benchmarks to reproduce 'merge' performance problems

StephanEwen committed:
[FLINK-4754] [checkpoints] Small followups to the configuration of number of retained checkpoints.

zentol committed:
[refactor] [py] General code clean-up

- added missing serialVersionUID
- PythonPlan* classes no longer implement serializable
- marked several fields as transient
- replaced some fields by local variables
- input/output files are now being deleted
- remvoed some unnecessary casts
- renamved several ignored exceptions
- close socket in PythonPlanStreamer
- simplified PyProcess health check
- eliminated several uses of raw types
- removed some unthrown exception declarations

aljoscha committed:
[hotfix] Remove auto-boxing in ContinuousEventTimeTrigger

StefanRRichter committed:
[hotfix] Added missing documentation in StateTransformationFunction

beyond1920 committed:
Introduce interface for external catalog, and provide an in-memory implementation for test or develop. Integrate with calcite catalog.

beyond1920 committed:
1. Update code to adapt to specify multiple scan packages in one tableSourceConverter.properties 2. Update code based on fabian's comment

beyond1920 committed:
1. add better comments on tableSourceProperties file. 2. add check for instantiable of scanned TableSourceConverter subclazz in ExternalTableSourceUtil

beyond1920 committed:
delete DataSchema, reuse tableSchema
